native/dlabrd.go

```diff
@@ -0,0 +1,143 @@
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlabrd reduces the first NB rows and columns of a real general m×n matrix
+// A to upper or lower bidiagonal form by an orthogonal transformation
+//  Q**T * A * P
+// If m >= n, A is reduced to upper bidiagonal form and upon exit the elements
+// on and below the diagonal in the first nb columns represent the elementary
+// reflectors, and the elements above the diagonal in the first nb rows represent
+// the matrix P. If m < n, A is reduced to lower bidiagonal form and the elements
+// P is instead stored above the diagonal.
+//
+// The reduction to bidiagonal form is stored in d and e, where d are the diagonal
+// elements, and e are the off-diagonal elements.
+//
+// The matrices Q and P are products of elementary reflectors
+//  Q = H_1 * H_2 * ... * H_nb
+//  P = G_1 * G_2 * ... * G_nb
+// where
+//  H_i = I - tauQ[i] * v_i * v_i^T
+//  G_i = I - tauP[i] * u_i * u_i^T
+//
+// As an example, on exit the entries of A when m = 6, n = 5, and nb = 2
+//  (  1   1   u1  u1  u1 )
+//  (  v1  1   1   u2  u2 )
+//  (  v1  v2  a   a   a  )
+//  (  v1  v2  a   a   a  )
+//  (  v1  v2  a   a   a  )
+// and when m = 5, n = 6, and nb = 2
+//  (  1   u1  u1  u1  u1  u1 )
+//  (  1   1   u2  u2  u2  u2 )
+//  (  v1  1   a   a   a   a  )
+//  (  v1  v2  a   a   a   a  )
+//  (  v1  v2  a   a   a   a  )
+//
+// Dlabrd also returns the matrices X and Y which are used with U and V to
+// apply the transformatino to the unreduced part of the matrix
+//  A := A - V*Y^T - X*U^T
+// and returns the matrices X and Y which are needed to apply the
+// transformation to the unreduced part of A.
+//
+// X is an m×nb matrix, Y is an n×nb matrix. d, e, taup, and tauq must all have
+// length at least nb. Dlabrd will panic if these size constraints are violated.
+func (impl Implementation) Dlabrd(m, n, nb int, a []float64, lda int, d, e, tauQ, tauP, x []float64, ldx int, y []float64, ldy int) {
+	checkMatrix(m, n, a, lda)
+	checkMatrix(m, nb, x, ldx)
+	checkMatrix(n, nb, y, ldy)
+	if len(d) < nb {
+		panic(badD)
+	}
+	if len(e) < nb {
+		panic(badE)
+	}
+	if len(tauQ) < nb {
+		panic(badTauQ)
+	}
+	if len(tauP) < nb {
+		panic(badTauP)
+	}
+	if m <= 0 || n <= 0 {
+		return
+	}
+	bi := blas64.Implementation()
+	if m >= n {
+		// Reduce to upper bidiagonal form.
+		for i := 0; i < nb; i++ {
+			bi.Dgemv(blas.NoTrans, m-i, i, -1, a[i*lda:], lda, y[i*ldy:], 1, 1, a[i*lda+i:], lda)
```

Comment:
Because of row vs. column major. Here a row of y is acting like a vector. Moving along a row has a stride of ldy in column major (like Lapack), but has a stride of 1 in row major (as in here).